elastic/kibana
# Fix telemetry plugin error crash

Resolves #122146.

The telemetry plugin can crash the Kibana server if it attempts to check for telemetry opt-in status while Elasticsearch doesn't have any shards available. This scenario seems to reliably occur when running functional tests locally.

This PR wraps the check for telemetry opt-in status in a try/catch, and returns `false` if there was an error.

[Per `get_telemetry_saved_object.ts`](https://github.com/jportner/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/src/plugins/telemetry/server/telemetry_repository/get_telemetry_saved_object.ts#L16-L35): a `null` result would indicate that the saved object was not found (and that Kibana could potentially opt-in), while a `false` result indicates that Kibana cannot opt-in.


### Testing

As mentioned in the linked PR, you can encounter this error while running a variety of functional tests.

To verify this fix solves the linked issue:

1. Start the functional test server (FTS) for a given suite, for example: `node scripts/functional_tests_server.js --config x-pack/test/saved_object_api_integration/spaces_only/config.ts`
2. Wait for the server to finish loading, you'll see a message like this:
   ```
   Elasticsearch and Kibana are ready for functional testing. Start the functional tests
   in another terminal session by running this command from this directory:
   
       node scripts/functional_test_runner --config x-pack/test/saved_object_api_integration/spaces_only/config.ts
   ```
3. Start the functional test runner (FTR) for the same suite, for example: `node scripts/functional_test_runner.js --config x-pack/test/saved_object_api_integration/spaces_only/config.ts`
4. Observe tests running normally 
5. Wait for a time, then observe the Elasticsearch error when Kibana tries to fetch telemetry opt-in status:
   ```
   │ info [r.suppressed] [node-01] path: /.kibana_8.1.0/_doc/telemetry%3Atelemetry, params: {index=.kibana_8.1.0, id=telemetry:telemetry}
   │      org.elasticsearch.action.NoShardAvailableActionException: No shard available for [get [.kibana_8.1.0][telemetry:telemetry]: routing [null]]
   ...
   ```
6. Observe that the FTS does _not_ crash, it continues to run

The telemetry plugin also writes a message to the debug logger, but that won't show up in the FTS logs by default.
